src/crimson/os/seastore/seastore_types.h

```diff
@@ -49,12 +49,23 @@ constexpr segment_id_t BLOCK_REL_SEG_ID =
 // for tests which generate fake paddrs
 constexpr segment_id_t FAKE_SEG_ID =
   std::numeric_limits<segment_id_t>::max() - 4;
+using device_id_t = uint8_t;
+constexpr uint16_t device_id_len = 4;
+// mask for segment manager id
+constexpr segment_id_t SM_ID_MASK =
+  0xF << (std::numeric_limits<segment_id_t>::digits - device_id_len);
+
+device_id_t get_device_id(segment_id_t id);
+segment_id_t add_device_id(segment_id_t id, device_id_t sm_id);
+segment_id_t strip_device_id(segment_id_t id);
```

Comment:
I'd like segment_id_t's to always have both the device id and the segment portions.  The only users of the raw segment are segment_cleaner and block.  Block can use separate type internally for its device_id-less segments (device_segment_t perhaps).  segment_cleaner probably needs to be converted to have a separate table for each device, and can use the block.h internal type in the mean time.  Perhaps convert segment_id_t to a struct with a single private uint32_t member and a constructor taking a device_id_t and a device_segment_t?

Reply:
I'll try that

Reply:
> I'd like segment_id_t's to always have both the device id and the segment portions.

Hmm.. I'm not sure hiding device-id into the segment-id is appropriate. If there are mutiple tiers of devices (not necessarily segmented devices), After getting the paddr from laddr, TransactionManager will need to dispatch reads according to the device-id, right?

Reply:
I don't want it to be easy to pass a segment_id_t without a device to an interface expecting it to have the device filled it.  In your scenario, TransactionManager will need to manipulate both kinds of ID, so it will be helpful for the type system to help us to keep them distinct.

Reply:
> Hmm.. I'm not sure hiding device-id into the segment-id is appropriate. If there are mutiple tiers of devices (not necessarily segmented devices), After getting the paddr from laddr, TransactionManager will need to dispatch reads according to the device-id, right?

TransactionManager doesn't have to do this, we can give EPM read functionality and redirect all reads to EPM, since EPM has the device_di<->device mapping.

Reply:
> I don't want it to be easy to pass a segment_id_t without a device to an interface expecting it to have the device filled it.

Maybe it's better to have `device_segment_id` for this case, and a different type `segment_id` for device_id-less segment logics.
My only point is that device_id is a generic portion of the paddr.

Reply:
Either way, I don't want segment_id_t to be used for both with and without the device bits.  I want them to be different types to enable the type system to help us to avoid bugs.